src/rgw/rgw_rest_s3.cc

```diff
@@ -6305,34 +6323,78 @@ int RGWSelectObj_ObjStore_S3::run_s3select(const char* query, const char* input,
       csv.use_header_info=true;
     }
 
-    m_s3_csv_object = std::unique_ptr<s3selectEngine::csv_object>(new s3selectEngine::csv_object(s3select_syntax.get(), csv));
+    if (m_parquet_type){
+      m_s3_parquet_object = std::unique_ptr<s3selectEngine::parquet_object>(new s3selectEngine::parquet_object(s->object->get_name(), s3select_syntax.get(), m_rgw_api));
+    } 
+    else {
+      m_s3_csv_object = std::unique_ptr<s3selectEngine::csv_object>(new s3selectEngine::csv_object(s3select_syntax.get(), csv));
+    }
   }
 
   header_size = create_header_records(m_buff_header.get());
   m_result.append(m_buff_header.get(), header_size);
   m_result.append(PAYLOAD_LINE);
 
-  if (s3select_syntax->get_error_description().empty() == false) {
+  std::function<int(std::string&)> fp_result_header_format = [&](std::string& result) {
+    
+    result = "012345678901";
+    header_size = create_header_records(m_buff_header.get());
+    result.append(m_buff_header.get(), header_size);
+    result.append(PAYLOAD_LINE);
+    return 0;
+  };
+
+  std::function<int(std::string&)> fp_s3select_result_format = [&](std::string &result) {
+
+    result.append(END_PAYLOAD_LINE);
+    int buff_len = create_message(result, result.size() - 12, header_size);
+    s->formatter->write_bin_data(result.data(), buff_len);
+    rgw_flush_formatter_and_reset(s, s->formatter);
+
+    return 0;
+  };
+
+  if (s3select_syntax->get_error_description().empty() == false)
+  {
     m_result.append(s3select_syntax->get_error_description());
     ldpp_dout(this, 10) << "s3-select query: failed to prase query; {" << s3select_syntax->get_error_description() << "}"<< dendl;
     status = -1;
   }
-  else {
-    status = m_s3_csv_object->run_s3select_on_stream(m_result, input, input_length, s->obj_size);
-    if(status<0) {
-      m_result.append(m_s3_csv_object->get_error_description());
+  else
+  {
+    if (m_parquet_type){
+            //TODO implement formatter call-back
+            status = m_s3_parquet_object->run_s3select_on_object(m_result,fp_s3select_result_format,fp_result_header_format);
+
+    }
+    else{
+            status = m_s3_csv_object->run_s3select_on_stream(m_result, input, input_length, s->obj_size);
+    }
+    if (status < 0){
+      if(m_parquet_type){
+      //m_result.append(m_s3_parquet_object->get_error_description()); //TODO get_error_description for runtime errors
+        ldout(s->cct, 10) << "s3select: failure while execution " << dendl;
+      } 
+      else {
+        m_result.append(m_s3_csv_object->get_error_description());
+      }
     }
   }
 
-  if (m_result.size() > strlen(PAYLOAD_LINE)) {
+  
+  //TODO for parquet processing should pass call-back to handle output
```

Comment:
if this is still TODO, should I assume the parquet engine actually doesn't work?

Reply:
the TODO comments are not relevant, will remove that.
parquet-handler uses callbacks (result-formatter and header-formatter)